internal/character/guinaifen/talent.go

```diff
@@ -0,0 +1,124 @@
+	Firekiss         = "guinaifen-firekiss"
+	E4Listener       = "guinaifen-e4-listener"
+)
+
+func init() {
+	modifier.Register(Talent, modifier.Config{
+		Listeners: modifier.Listeners{
+			OnBeforeAttack: A2,
+		},
+	})
+
+	modifier.Register(FirekissListener, modifier.Config{
+		Listeners: modifier.Listeners{
+			OnBeforeBeingHitAll: checkFirekiss,
+		},
+	})
+
+	modifier.Register(Firekiss, modifier.Config{
+		StatusType: model.StatusType_STATUS_DEBUFF,
+		Stacking:   modifier.ReplaceBySource,
+		Listeners: modifier.Listeners{
+			OnAdd: FirekissonStack,
+		},
+	})
+}
+
+func (c *char) initTalent() {
+	c.engine.AddModifier(c.id, info.Modifier{
+		Name:   Talent,
+		Source: c.id,
+	})
+
+	// apply Firekiss listener to all enemies at start
+	for _, trg := range c.engine.Enemies() {
+		c.engine.Events().CharactersAdded.Subscribe(func(e event.CharactersAdded) {
+			c.engine.AddModifier(trg, info.Modifier{
+				Name:   FirekissListener,
+				Source: c.id,
+			})
+		})
+	}
+
+	// apply E4 listener to all enemies at the start
+	if c.info.Eidolon >= 4 {
+		for _, trg := range c.engine.Enemies() {
+			c.engine.Events().CharactersAdded.Subscribe(func(e event.CharactersAdded) {
+				c.engine.AddModifier(trg, info.Modifier{
+					Name:   E4Listener,
+					Source: c.id,
+				})
+			})
+		}
+	}
+}
+
+func checkFirekiss(mod *modifier.Instance, e event.HitStart) {
+
+	gui, _ := mod.Engine().CharacterInfo(mod.Source())
+	maxStacks := 3
+	if gui.Eidolon >= 6 {
+		maxStacks = 4
+	}
+
+	// check if damage was from Burn using workaround
+	if e.Hit.AttackType == model.AttackType_DOT && e.Hit.DamageType == model.DamageType_FIRE {
+		mod.Engine().AddModifier(e.Defender, info.Modifier{
+			Name:              Firekiss,
+			Source:            mod.Source(),
+			Chance:            1,
+			Duration:          3,
+			MaxCount:          float64(maxStacks),
+			CountAddWhenStack: 1,
+		})
+	}
+}
+
+// calculate the received damage increase
+func FirekissonStack(mod *modifier.Instance) {
+	gui, _ := mod.Engine().CharacterInfo(mod.Source())
+	mod.SetProperty(prop.AllDamageTaken, mod.Count()*talent[gui.TalentLevelIndex()])
+}
+
+// apply DoT on Normal
+func A2(mod *modifier.Instance, e event.AttackStart) {
+	gui, _ := mod.Engine().CharacterInfo(mod.Owner())
+	if gui.Traces["101"] && e.AttackType == model.AttackType_NORMAL {
+		target := e.Targets[0]
+		if gui.Eidolon >= 2 && mod.Engine().HasBehaviorFlag(target, model.BehaviorFlag_STAT_DOT_BURN) {
+			mod.Engine().AddModifier(target, info.Modifier{
+				Name:   common.Burn,
+				Source: mod.Owner(),
+				State: &common.BurnState{
+					DamagePercentage: skillBurn[gui.SkillLevelIndex()] + 0.4,
+				},
+				Chance:   0.8,
+				Duration: 2,
+			})
+		} else {
+			mod.Engine().AddModifier(target, info.Modifier{
+				Name:   common.Burn,
+				Source: mod.Owner(),
+				State: &common.BurnState{
+					DamagePercentage: skillBurn[gui.SkillLevelIndex()],
+				},
+				Chance:   0.8,
+				Duration: 2,
+			})
+		}
```

Comment:
same as comment on skill, could technically extract the actual burn application into a separate function and have both of them call it (with a parameter for the base chance to use)